app/controllers/api/v1/user_annotations_controller.rb

```diff
@@ -0,0 +1,119 @@
+      include Concerns::StudyAware
+      include Swagger::Blocks
+
+      before_action :set_current_api_user!
+      before_action :set_study
+      before_action :check_study_view_permission
+
+      # Define parameters used in API POST endpoint to create user annotations
+      swagger_schema :UserAnnotationInput do
+        key :name, 'UserAnnotation'
+        property :name do
+          key :type, :string
+          key :description, 'Name of new custom user annotation'
+        end
+        property :user_data_arrays_attributes do
+          key :type, :object
+          key :description, 'Labels (names) and cell name arrays (values) for each selection'
+        end
+        property :cluster do
+          key :type, :string
+          key :description, 'Cluster group name'
+        end
+        property :subsample_threshold do
+          key :type, :string
+          key :description, 'Subsample threshold'
+        end
+        property :subsample_annotation do
+          key :type, :string
+          key :description, 'Subsample annotation'
+        end
+      end
+
+      swagger_path '/studies/{accession}/user_annotation' do
+        operation :post do
+          key :tags, [
+              'Studies'
+          ]
+          key :summary, 'Create user annotation'
+          key :description, 'Create new custom user annotation for the study'
+          key :operationId, 'site_study_user_create_annotation_path'
+          parameter do
+            key :name, :accession
+            key :in, :path
+            key :description, 'Study accession number (e.g. SCP123)'
+            key :required, true
+            key :type, :string
+          end
+          parameter do
+            key :name, :user_annotation
+            key :in, :body
+            key :description, 'UserAnnotationInput object'
+            schema do
+              key :'$ref', :UserAnnotationInput
+            end
+          end
+          response 200 do
+            key :description, 'Successfully created user annotation'
+          end
+          extend SwaggerResponses::StudyControllerResponses
+        end
+      end
+
+      def create
+
+        begin
+          # Parameters to log for any errors
+          user_annotation_params = params.dup
+          user_annotation_params[:current_user] = current_user
+          log_params = user_annotation_params.dup
+          # Don't log data_arrays; it's too big
+          log_params.delete(:user_data_arrays_attributes)
+
+          cluster = params[:cluster]
+
+          message, annotations = UserAnnotationService.create_user_annotation(
+            @study, params[:name], params[:user_data_arrays_attributes],
+            cluster, params[:loaded_annotation],
+            params[:subsample_threshold], current_api_user
+          )
+
+          render json: {message: message, annotations: annotations}
+
+        # Handle other errors in saving user annotation
+        rescue Mongoid::Errors::InvalidValue => e
```

Comment:
I think I was too vague in my earlier comments about where I think this excpetion handling code belongs.  I LOVE this specific exception catching, but it belongs in the service method.  This controller shouldn't need to know anything about Mongoid-specific errors (or that the service uses Mongoid at all).  So this `rescue Mongoid::Errors::InvalidValue` block should be in the service, and it should handle the exception by re-throwing it as an ArgumentError.  Then the more general rescue block below (which should stay in the controller) will handle rendering it to the client.

Reply:
Done in 655b981.